vent/io/devices/sensors.py

```diff
@@ -14,76 +15,85 @@ class Sensor(ABC):
+        self._data = {
+            'timestamp': deque(maxlen=self.maxlen_data),
+            'value': deque(maxlen=self.maxlen_data)
+        }
 
     def update(self) -> float:
         """ Make a sensor reading, verify that it makes sense and store the result internally. Returns True if reading
         was verified and False if something went wrong.
         """
         value = self._read()
-        if self._verify(value):
-            self.__store_last(value)
-            self._last_timestamp = time.time()
+        self._data['timestamp'].append(time.time())
+        self._data['value'].append(value)
         return self._verify(value)
 
-    def get(self) -> float:
-        """ Return the most recent sensor reading."""
-        if self._last_timestamp == -1:
-            raise RuntimeWarning('get() called before update()')
-        return self._data[(self._i - 1) % self._data_length]
+    def get(self, average=False) -> float:
+        """ Return the most recent sensor reading, or an average of readings since last get(). Clears internal memory so as not to have stale data."""
+        if len(self._data['value']) == 0:
+            self.update()
+        if average:
+            value = sum(self._data['value'])/len(self._data['value'])
+        else:
+            value = self._data['value'].pop()
+        self._clear()
+        return value
 
     def age(self) -> float:
         """ Returns the time in seconds since the last sensor update, or -1 if never updated."""
-        if self._last_timestamp == -1:
+        if not self._data['timestamp']:
             return -1.0
         else:
-            return time.time() - self._last_timestamp
+            return time.time() - self._data['timestamp'][-1]
 
     def reset(self):
         """ Resets the sensors internal memory. May be overloaded by subclasses to extend functionality specific to a
         device.
         """
-        self._data = np.zeros(self.data_length, dtype=np.float16)
-        self._i = 0
+        self._clear()
+            
+    def _clear(self):
+        """ Resets the sensors internal memory. """
+        for key in self._data.keys():
+            self._data[key].clear()
 
     @property
-    def data(self) -> np.ndarray:
-        """ Locally-stored observations.
-        Note: ndarray.astype(bool) returns an equivalent sized array
-        with True for each nonzero element and False everywhere else.
+    def data(self) -> np.array:
+        """ Returns all Locally-stored observations. 
 
         Returns:
-            np.ndarray: an ndarray of observations arranged oldest to newest. Result has length equal to the lessor
-                of self.n and the number of observations.
-        made.
+            np.array: An array of timestamped observations arranged oldest to newest.
         """
-        rolled = np.roll(self._data, self.data_length - self._i)
-        return rolled[rolled.astype(bool)]
+        result = np.array(

```

Comment:
```python
np.column_stack([self._data['timestamp'], self._data['value'])
```
is roughly 3x as fast for dqs of appreciable length

Reply:
sweet